client/app/bundles/course/assessment/submission/actions/index.js

```diff
@@ -484,15 +485,70 @@ export function saveGrade(submissionId, grades, exp, published) {
   };
 
   return (dispatch) => {
-    dispatch({ type: actionTypes.SAVE_GRADE_REQUEST });
+    dispatch({ type: actionTypes.SAVE_ALL_GRADE_REQUEST });
 
     return CourseAPI.assessment.submissions
       .update(submissionId, payload)
       .then((response) => response.data)
       .then((data) => {
-        dispatch({ type: actionTypes.SAVE_GRADE_SUCCESS, payload: data });
+        dispatch({ type: actionTypes.SAVE_ALL_GRADE_SUCCESS, payload: data });
         dispatch(setNotification(translations.updateSuccess));
       })
+      .catch((error) => {
+        dispatch({ type: actionTypes.SAVE_ALL_GRADE_FAILURE });
+        dispatch(
+          setNotification(translations.updateFailure, buildErrorMessage(error)),
+        );
+      });
+  };
+}
+
+export function saveGrade(
+  submissionId,
+  grades,
+  allGrades,
+  exp,
+  published,
+  questionTitle,
+) {
+  const expParam = published ? 'points_awarded' : 'draft_points_awarded';
+  const payload = {
+    submission: {
+      answers: grades,
+      [expParam]: exp,
+    },
+  };
+
+  const mapAnswerIdToAnswer = arrayToObjectById(allGrades);
+
+  return (dispatch) => {
+    dispatch({ type: actionTypes.SAVE_GRADE_REQUEST });
+
+    return CourseAPI.assessment.submissions
+      .update(submissionId, payload)
+      .then((response) => response.data)
+      .then((data) => {
+        const dataWithPrefilledGrades = {
```

Comment:
use `produce` from `immer`.

Reply:
Also, update only the grade that is saved to the backend. This function took the grade of 1 question. To avoid weird unexpected behaviour, it should only update the grade of that particular question in redux.